File: fibratus/output/console.py
# Copyright 2016 by Nedim Sabic (RabbitStack)
# All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.
import json
from _ctypes import byref

from fibratus.apidefs.sys import get_std_handle, STD_OUTPUT_HANDLE, write_console_unicode, c_ulong
from fibratus.output.base import Output

RENDER_FORMAT = '%s %s %s %s (%s) - %s %s'


class ConsoleOutput(Output):

    def __init__(self, **kwargs):
        Output.__init__(self)

        self._fmt = kwargs.pop('format', 'pretty')
        self._timestamp_pattern = kwargs.pop('timestamp_pattern', '%Y-%m-%d %H:%M:%S.%f')
        self._stdout_handle = get_std_handle(STD_OUTPUT_HANDLE)

        assert self._stdout_handle, 'could not acquire the standard output stream handle'

    def emit(self, kevent, **kwargs):
        """Renders the kevent to the standard output stream.

        Uses the default output format or JSON to render the
        kernel event to standard output stream.

        The default output format is as follows:

        id  timestamp  cpu  process  (process id) - kevent (parameters)
        --  ---------  ---  -------  -----------   ------- ------------

        Example:

        160 13:27:27.554 0 wmiprvse.exe (1012) - CloseFile (file=C:\\WINDOWS\\SYSTEM32\\RSAENH.DLL, tid=2668)

        Parameters
        ----------

        kevent: KEvent
            the information regarding the kernel event

        kwargs: dict
            console adapter configuration

        """
        pid, proc = kevent.get_thread()
        if 'pretty' in self._fmt:
            kevt = RENDER_FORMAT % (kevent.kid,
                                    kevent.ts.time(),
                                    kevent.cpuid,
                                    proc,
                                    pid,
                                    kevent.name,
                                    self._format_params(kevent.params))
        else:
            kevt = json.dumps(dict(id=kevent.kid,
                                   timestamp=kevent.ts.strftime(self._timestamp_pattern),
                                   cpuid=kevent.cpuid,
                                   proc=proc,
                                   pid=pid,
                                   name=kevent.name,
                                   params=kevent.params))

        kevt += '\n'
        # write the output on the standard output stream
        write_console_unicode(self._stdout_handle, kevt,
                              len(kevt),
                              byref(c_ulong()),
                              None)

    def _format_params(self, kparams):
        """Transforms the kevent parameters.

        Apply the rendering format on the kevent payload
        to transform it into more convenient structure
        sorted by parameter keys.
        """
        fmt = ', '.join('%s=%s' % (k, kparams[k]) for k in sorted(kparams.keys())) \
            .replace('\"', '')
        return '(%s)' % fmt

File: pkg/outputs/console/console.go
/*
 * Copyright 2019-2020 by Nedim Sabic Sabic
 * https://www.fibratus.io
 * All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package console

import (
	"bufio"
	"expvar"
	"github.com/rabbitstack/fibratus/pkg/kevent"
	"github.com/rabbitstack/fibratus/pkg/outputs"
	"os"
)

var (
	consoleErrors = expvar.NewInt("output.console.errors")
)

type format string

const (
	pretty format = "pretty"
	json   format = "json"
	// template represents the default template used in pretty rendering mode
	template = "{{ .Seq }} {{ .Timestamp }} - {{ .CPU }} {{ .Process }} ({{ .Pid }}) - {{ .Type }} ({{ .Kparams }})"
)

type console struct {
	writer    *bufio.Writer
	formatter *kevent.Formatter
	format    format
}

func init() {
	outputs.Register(outputs.Console, initConsole)
}

func initConsole(config outputs.Config) (outputs.OutputGroup, error) {
	stdout := os.Stdout
	cfg, ok := config.Output.(Config)
	if !ok {
		return outputs.Fail(outputs.ErrInvalidConfig(outputs.Console, config.Output))
	}
	tmpl := cfg.Template
	if tmpl == "" {
		tmpl = template
	}
	formatter, err := kevent.NewFormatter(tmpl)
	if err != nil {
		return outputs.Fail(err)
	}
	if cfg.ParamKVDelimiter != "" {
		kevent.ParamKVDelimiter = cfg.ParamKVDelimiter
	}

	c := &console{
		writer:    bufio.NewWriterSize(stdout, 8*1024),
		formatter: formatter,
		format:    format(cfg.Format),
	}
	return outputs.Success(c), nil
}

func (c *console) Close() error   { return c.writer.Flush() }
func (c *console) Connect() error { return nil }
func (c *console) Publish(batch *kevent.Batch) error {
	for _, kevt := range batch.Events {
		var buf []byte
		switch c.format {
		case json:
			buf = kevt.MarshalJSON()
		case pretty:
			buf = c.formatter.Format(kevt)
		default:
			return nil
		}

		if err := c.write(buf); err != nil {
			consoleErrors.Add(1)
			continue
		}
		if err := c.write(nl); err != nil {
			consoleErrors.Add(1)
			continue
		}
	}

	if err := c.writer.Flush(); err != nil {
		consoleErrors.Add(1)
		return err
	}

	return nil
}

var nl = []byte("\n")

func (c *console) write(buf []byte) error {
	written := 0
	for written < len(buf) {
		n, err := c.writer.Write(buf[written:])
		if err != nil {
			return err
		}
		written += n
	}
	return nil
}
